driver-legacy/src/main/com/mongodb/LegacyMixedBulkWriteOperation.java

```diff
@@ -47,64 +40,42 @@
 
 
 /**
- * Abstract base class for write operations.
- *
- * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ * Abstract base class for legacy write operations.
  */
-public abstract class BaseWriteOperation implements AsyncWriteOperation<WriteConcernResult>, WriteOperation<WriteConcernResult> {
+class LegacyMixedBulkWriteOperation implements WriteOperation<WriteConcernResult> {
     private final WriteConcern writeConcern;
     private final MongoNamespace namespace;
+    private final List<? extends WriteRequest> writeRequests;
+    private final WriteRequest.Type type;
     private final boolean ordered;
     private final boolean retryWrites;
     private Boolean bypassDocumentValidation;
-    private BsonValue comment;
 
-    public BaseWriteOperation(final MongoNamespace namespace, final boolean ordered, final WriteConcern writeConcern,
-                              final boolean retryWrites) {
+    LegacyMixedBulkWriteOperation(final MongoNamespace namespace, final boolean ordered, final WriteConcern writeConcern,
+            final boolean retryWrites, final List<? extends WriteRequest> writeRequests) {
+        this.writeRequests = notNull("writeRequests", writeRequests);
+        isTrueArgument("writeRequests not empty", !writeRequests.isEmpty());
+        this.type = writeRequests.get(0).getType();
```

Comment:
There appears to be nothing preventing `writeRequests` from containing requests of different types. We can change this by making the constructor `private` and exposing `static` methods that take lists of specific subtypes of `WriteRequest` that are not subtypes of each other.

Reply:
Done, though since update and replace are both represented by UpdateRequest, I didn't bother to try to distinguish between them.  We could add an assertion to ensure that update and replace are not mixed, but seems like overkill.